Bearded.Graphics/Debugging/KHRDebugExtension.cs

```diff
@@ -0,0 +1,57 @@
+using System.Linq;
+using OpenTK.Graphics.OpenGL;
+
+namespace Bearded.Graphics.Debugging
+{
+    public abstract class KHRDebugExtension
+    {
+        private static KHRDebugExtension? instance;
+        public static KHRDebugExtension Instance => instance ??= createInstanceWithAvailableFunctionality();
+
+        private KHRDebugExtension()
+        {
+        }
+
+        private static KHRDebugExtension createInstanceWithAvailableFunctionality()
+        {
+            if (khrDebugExtensionIsAvailable())
+                return new ActualImplementation();
+            return new DummyImplementation();
+        }
+
+        private static bool khrDebugExtensionIsAvailable()
+        {
+            var count = GL.GetInteger(GetPName.NumExtensions);
+
+            return Enumerable.Range(0, count)
+                .Select(i => GL.GetString(StringNameIndexed.Extensions, i))
+                .Contains("GL_KHR_debug");
+        }
+
+        public virtual void SetObjectLabel(ObjectLabelIdentifier identifier, int name, string label) { }
```

Comment:
Isn't the "not doing anything" aspect here an implementation detail of the dummy implementation? Making these methods abstract would be more expected here. There is nothing preventing anybody from creating their own implementation (though why would they), and having it default to not doing anything on these methods sounds error prone. This applies even if we are adding new methods here: there is no compiler checking that the actual implementation actually ends up implementing these methods.

In this instance, I would choose to accept the extra boilerplate, make these methods abstract, and make the dummy implementation explicit in implementing these as empty methods.

Reply:
> There is nothing preventing anybody from creating their own implementation

Yes, there is: the private constructor. :)

Given that, and that this currently saves a couple of lines of code, I'd say it's fine this way. If it really bothers you, I can make them abstract of course.